### ISSUE TITLE
Deploy Triton server with MinIO as Model Store

### ISSUE THREAD

**Issue by chiehpower** [triton-inference-server/server#4102]
**Description**
Hi all,
I was trying to deploy the Tirton server with MinIO as a model store. 
I have referred to this article; however, this way is to deploy on AWS S3. 

My Triton server and MinIO are all deployed on localhost. (In my personal desktop)

I already uploaded the model to MinIO. (The model was downloaded from [here](https://github.com/triton-inference-server/server/blob/main/docs/examples/fetch_models.sh))

![screenshot-10-1-2-84-9001-buckets-models-browse-1648101226130](https://user-images.githubusercontent.com/32332200/159851507-a937d6b4-9b21-4a9b-9d94-286ec6e7bb4f.png)

I am wondering what location should I need to put in the option of `--model-store=`?

Here is my error message, but obviously the location of `model-store` is not correct. 
![image](https://user-images.githubusercontent.com/32332200/159852668-7da2353f-b091-476d-a350-8a613c05002e.png)

**Triton Information**
What version of Triton are you using? `nvcr.io/nvidia/tritonserver:21.03-py3`

Are you using the Triton container or did you build it yourself?
Triton container from NGC

**To Reproduce**
Steps to reproduce the behavior.

1. Start a minIO server
```
docker run -p 9000:9000 -p 9001:9001 --name minio1 \
                -v data:/data \
                -v config:/root/.minio \
                minio/minio:RELEASE.2022-02-01T18-00-14Z \
                server /data --console-address "0.0.0.0:9001"
```

Create a bucket, and upload the model to server.

2. Start a Triton as usual, but you need to fill in the location for  `--model-store`

**Expected behavior**
The models stored on minIO can be loaded from Triton inference server.

If is there any part that I elaborated not clear, please let me know.
Thank you!

BR,
Chieh

**Comment by bezero** [comment]
If you run trion server docker with `--network host` you will be able to use localhost to refer to your local minio installation. 
An even better solution would be if you use docker-compose to start both of your services in the same network.

```yaml
# sample docker cmpose file
version: '3.4'

services:
  minio:
    image: minio/minio:RELEASE.2022-02-01T18-00-14Z
    ports:
      - 9000:9000
      - 9001:9001
    volumes:
      - ./data:/data"
      - ./config:/root/.minio
    environment:
      MINIO_ACCESS_KEY: ${AWS_ACCESS_KEY_ID}
      MINIO_SECRET_KEY: ${AWS_SECRET_ACCESS_KEY}
      S3_TRITON_BUCKET: ${S3_TRITON_BUCKET}
    entrypoint: sh
    command: -c 'mkdir -p /data/$S3_TRITON_BUCKET && server /data'
    restart: unless-stopped

  inference_server:
    image: "nvcr.io/nvidia/tritonserver:21.03-py3"
    restart: unless-stopped
    ports:
      - 8000:8000
      - 8002:8002
    expose:
      - 8000
      - 8002  
    command: ["tritonserver", "--strict-model-config=false", "--model-control-mode=explicit", "--model-store=s3://minio:9000/${S3_TRITON_BUCKET}"]
    shm_size: 1g
    ulimits:
      memlock: -1
      stack: 67108864
```

**Comment by chiehpower** [comment]
Hi @bezero ,

Thank you for your reply! 
I appreciate your sample file!

My problem seems not because of the container network issue. (so change to host network not helping) 
Indeed, I wrote it in the docker-compose file as well. (For the reason of easier reproducing, I did not write the way of docker-compose in the beginning.) 

However, I saw your location which was still using s3:// and the variable (i.e., S3_TRITON_BUCKET) of the environment area (According to my understanding, it should be relied on AWS, correct me if I figure out wrong please)

**My scenario is not using any cloud service (storage).** All services are deployed on the same desktop. 

If using your sample here, I have nothing to fill in the environment area of minio service. And at the command item, I will change to the normal launching command without creating any folder. 

```
# sample docker compose file
version: '3.4'

services:
  minio:
    image: minio/minio:RELEASE.2022-02-01T18-00-14Z
    ports:
      - 9000:9000
      - 9001:9001
    volumes:
      - ./data:/data"
      - ./config:/root/.minio
    entrypoint: sh
    command: server /data --console-address "0.0.0.0:9001"
    restart: unless-stopped

  inference_server:
    image: "nvcr.io/nvidia/tritonserver:21.03-py3"
    restart: unless-stopped
    ports:
      - 8000:8000
      - 8002:8002
    expose:
      - 8000
      - 8002  
    command: ["tritonserver", "--strict-model-config=false", "--model-control-mode=explicit", "--model-store=(what should I fill in here?)"]
    shm_size: 1g
    ulimits:
      memlock: -1
      stack: 67108864
```

And then what should I fill in the information in the argument of `model-store`?

Thank you so much!!
BR,
Chieh

**Comment by bezero** [comment]
@chiehpower 
In your case, assuming you have created a bucket let's say `triton` and you have a models folder in that bucket as `triton/models/`
Then, your path should be ` "--model-store=s3://minio:9000/triton/models"`
Also, don't forget to pass minio credentials to your triton service (which I did forget to include in my previous response)
```yaml
# sample docker compose file
services:
  inference_server:
    image: "nvcr.io/nvidia/tritonserver:21.03-py3"
    restart: unless-stopped
    environment:
      - OMP_NUM_THREADS=4
      - AWS_ACCESS_KEY_ID=${AWS_ACCESS_KEY_ID}
      - AWS_SECRET_ACCESS_KEY=${AWS_SECRET_ACCESS_KEY}
      - AWS_DEFAULT_REGION=${AWS_DEFAULT_REGION}
    ports:
      - 8000:8000
      - 8002:8002
    expose:
      - 8000
      - 8002  
    command: ["tritonserver", "--strict-model-config=false", "--model-control-mode=explicit", "--model-store=s3://minio:9000/triton/models"]
    shm_size: 1g
    ulimits:
      memlock: -1
      stack: 67108864
```

**Comment by chiehpower** [comment]
Hi @bezero 

Thanks for your detailed explanation.

After I did several experiments and testings, eventually I could successfully make minIO with Triton server work well.
Also, I have tested that it can do inference well from Triton.

Thanks again!!

Attachment: for someone who wanna try it directly, the code (script) is [here](https://github.com/chiehpower/Setup-deeplearning-tools/blob/master/minIO/docker-compose-triton.yml).

BR,
Chieh
